authentication/authentication.go

```diff
@@ -214,3 +224,127 @@ func (a *Authentication) UserInfo(ctx context.Context, accessToken string, opts
 	err = a.Request(ctx, "GET", a.URI("userinfo"), nil, &user, opts...)
 	return
 }
+
+// Helper for adding values to a url.Values instance if they are not empty.
+func addIfNotEmpty(key string, value string, qs url.Values) {
+	if value != "" {
+		qs.Set(key, value)
+	}
+}
+
+// Helper for enforcing that required values are set.
+func check(errors *[]string, key string, c bool) {
+	if !c {
+		*errors = append(*errors, key)
+	}
+}
+
+// Helper for adding client authentication into a url.Values instance.
+func (a *Authentication) addClientAuthenticationToURLValues(params oauth.ClientAuthentication, body url.Values, required bool) error {
+	clientID := params.ClientID
+	if params.ClientID == "" {
+		clientID = a.clientID
+	}
+	body.Set("client_id", clientID)
+
+	clientSecret := params.ClientSecret
+	if params.ClientSecret == "" {
+		clientSecret = a.clientSecret
+	}
+
+	switch {
+	case a.clientAssertionSigningKey != "" && a.clientAssertionSigningAlg != "":
+		clientAssertion, err := createClientAssertion(
+			a.clientAssertionSigningAlg,
+			a.clientAssertionSigningKey,
+			clientID,
+			a.url.JoinPath("/").String(),
+		)
+		if err != nil {
+			return err
+		}
+
+		body.Set("client_assertion", clientAssertion)
+		body.Set("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+		break
+	case params.ClientAssertion != "":
+		body.Set("client_assertion", params.ClientAssertion)
+		body.Set("client_assertion_type", params.ClientAssertionType)
```

Comment:
Should we check that `params.ClientAssertionType` is not empty? Like e.g. 

```go
case params.ClientAssertion != "" && params.ClientAssertionType != ""
```

Reply:
Yeah that makes sense, will fix

Reply:
Added in 716bdadcfdeeafb9a3bd86d11f384a8d7937a59e and also took the opportunity to extend `addClientAuthenticationToClientAuthStruct` to allow specifying if a client secret/client assertion is required